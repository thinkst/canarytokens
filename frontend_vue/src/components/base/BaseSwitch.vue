<!-- eslint-disable vuejs-accessibility/label-has-for -->
<template>
  <div class="relative flex flex-col justify-between">
    <input
      v-bind="$attrs"
      :id="id"
      type="checkbox"
      role="switch"
      class="toggle"
      :value="modelValue"
      :checked="modelValue"
      :aria-checked="modelValue"
      @change="handleToggle"
    />
    <label
      :for="id"
      class="relative"
      :class="[
        { multiline: helperMessage || errorMessage },
        { loading: loading },
      ]"
      >{{ label }}
      <BaseSpinner
        v-if="loading"
        class="absolute right-[0.6rem]"
        :class="[helperMessage ? 'top-[0.7rem]' : 'top-[0.2rem]']"
        height="1rem"
        :variant="modelValue === true ? 'secondary' : 'primary'"
      ></BaseSpinner>
    </label>
    <div>
      <p
        v-show="helperMessage"
        id="helper"
        class="text-xs leading-4 text-grey-500 pr-[3rem]"
      >
        {{ helperMessage }}
      </p>
      <p
        v-show="hasError"
        id="error"
        class="text-xs leading-4 text-red"
      >
        {{ errorMessage }}
      </p>
    </div>
  </div>
</template>

<script setup lang="ts">
defineProps<{
  id: string;
  label: string;
  helperMessage?: string | null;
  hasError?: boolean;
  errorMessage?: string;
  loading?: boolean;
  modelValue?: boolean;
}>();

const emit = defineEmits(['update:modelValue']);

function handleToggle(event: Event) {
  emit('update:modelValue', (event.target as HTMLInputElement).checked);
}
</script>

<style scoped>
input[type='checkbox'].toggle {
  opacity: 0;
  position: absolute;
  left: 0px;
  top: 0px;
}

label {
  position: relative;
  display: flex;
  align-items: center;
}

label:not(.loading) {
  cursor: pointer;
}

label:is(.loading) {
  pointer-events: none;
}

/* toggle wrapper */
label::before {
  content: '';
  position: absolute;
  right: 0;
  width: 2.2em;
  height: 1.4em;
  background-color: hsl(0, 0%, 91%);
  border-radius: 1em;
  transition: background-color 200ms ease-in-out;
}

label.multiline::before {
  top: 0.5rem;
}

label.multiline::after {
  top: 0.75rem;
}

/* toggle ball */
label:not(.loading)::after {
  content: '';
  display: flex;
  position: absolute;
  font-size: 0.5em;
  right: 2.2em;
  width: 1.8em;
  height: 1.8em;
  background-color: hsl(0, 0%, 70%);
  border-radius: 1em;
  transition:
    background-color 150ms ease-in-out,
    transform 100ms ease-in-out,
    height 100ms ease-in-out,
    width 100ms ease-in-out;
}

input[type='checkbox'].toggle:checked + label::before {
  background-color: hsl(152, 59%, 48%);
}

input[type='checkbox'].toggle:checked + label::after {
  transform: translateX(1.9em);
  background-color: white;
}

input[type='checkbox'].toggle:enabled:focus + label::after {
  outline: 0.8em solid hsla(0, 0%, 43%, 0.2);
}

input[type='checkbox'].toggle:enabled:active + label::after {
  outline: 0.8em solid hsla(0, 0%, 43%, 0.2);
  width: 2.1em;
  height: 2.1em;
}

input[type='checkbox'].toggle:checked:disabled + label::before {
  background-color: hsl(152, 59%, 73%);
}

input[type='checkbox'].toggle:disabled + label::before {
  background-color: hsl(0, 0%, 82%);
}

input[type='checkbox'].toggle:disabled + label::after {
  background-color: hsl(0, 0%, 70%);
}

input[type='checkbox'].toggle:disabled + label,
input[type='checkbox'].toggle:checked:disabled + label {
  pointer-events: none;
}
</style>
