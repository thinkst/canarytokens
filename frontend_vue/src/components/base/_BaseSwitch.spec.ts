import { mount } from '@vue/test-utils';
import BaseSwitch from './_BaseSwitch.vue';

describe('BaseSwitch.vue', () => {

  test('is called', () => {
    const wrapper = mount(BaseSwitch);
    expect(wrapper.exists()).toBeTruthy();
  });

  test('renders correctly', () => {
    const label = 'Test Label';
    const id = 'test-id';
    const wrapper = mount(BaseSwitch, {
      props: { label, id },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders label when passed', () => {
    const label = 'Test Label';
    const id = 'test-id';
    const wrapper = mount(BaseSwitch, {
      props: { label, id },
    });
    expect(wrapper.text()).toMatch(label);
  });

  it('emits an update:modelValue event when checked', async () => {
    const label = 'Test Label';
    const id = 'test-id';
    const wrapper = mount(BaseSwitch, {
      props: { label, id, modelValue: false },
    });
    const checkbox = wrapper.find('input[type="checkbox"]');
    //@ts-ignore
    await checkbox.setChecked(true);
    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([true]);
  });

  it('emits an update:modelValue event when unchecked', async () => {
    const label = 'Test Label';
    const id = 'test-id';
    const wrapper = mount(BaseSwitch, {
      props: { label, id, modelValue: true },
    });
    const checkbox = wrapper.find('input[type="checkbox"]');
    //@ts-ignore
    await checkbox.setChecked(false);
    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([false]);
  });

  it('renders a disabled checkbox when disabled is true', () => {
    const label = 'Test Label';
    const id = 'test-id';
    const wrapper = mount(BaseSwitch, {
      props: { label, id, disabled: true },
    });
    const checkbox = wrapper.find('input[type="checkbox"]');
    expect(checkbox.attributes('disabled')).toBeDefined();
  });
});
