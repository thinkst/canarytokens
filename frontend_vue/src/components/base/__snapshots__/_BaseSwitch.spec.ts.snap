// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseSwitch.vue > render correctly 1`] = `"<div data-v-c3331bcf="" class="flex flex-col-reverse justify-between"><input data-v-c3331bcf="" id="test-id" type="checkbox" role="switch" class="toggle" aria-checked="false"><label data-v-c3331bcf="" for="test-id">Test Label</label></div>"`;

exports[`BaseSwitch.vue > renders correctly 1`] = `"<div data-v-c3331bcf="" class="flex flex-col-reverse justify-between"><input data-v-c3331bcf="" id="test-id" type="checkbox" role="switch" class="toggle" aria-checked="false"><label data-v-c3331bcf="" for="test-id">Test Label</label></div>"`;
